### ISSUE TITLE
Initialization of action constructor when mocked

### ISSUE THREAD

**Issue by BoyeMagnus** [lorisleiva/laravel-actions#118]
**Steps to reproduce:**

1. Create a action with a dependency injection.
2. Mock the action, and set the mock to pass though (to the handle function).
3. The dependency injection is not injected.

**Example:**
When mocking an action the dependencies injected through the constructor does not seem to be available, when allowing passthrus.

I'm trying to mock the following action:
![image](https://user-images.githubusercontent.com/24491343/119716152-19dfbe00-be65-11eb-842c-11973f4af0d1.png)

Which depends on the RangeKeyService (which is a singleton in Laravel). 
![image](https://user-images.githubusercontent.com/24491343/119716191-25cb8000-be65-11eb-8d64-d7f2acde8eaf.png)

The constructor is correctly called when doing `GetContextValueObjectFromUser::mock()`, where I would expect it to set the RangeKeyService <---- **However, it doesn't seem to be set when running the mocked action later. Here the $this->rangeKeyService is null.**

The constructor is correctly NOT called when doing `GetContextValueObjectFromUser::partialMock()`, where I wouldn't expect it to set the RangeKeyService

The error thrown is the following:
Error: Typed property GetContextValueObjectFromUser::$rangeKeyService must not be accessed before initialization

Should this be working as I'm expecting or might I be wrong regarding something with mocking?

**Comment by lorisleiva** [comment]
Hi there 👋 

Sorry for the late reply. I've done some digging and here are my findings.

When mocking using `mock` or `partialMock` it doesn't call the constructor. If you `dd` into the constructor you will successfully see the content of that `dd` but that's because Laravel Actions must instantiate an instance before deciding if it's a fake or not.

When mocking using `passthru`, the return value will use the real function but in a very limited fashion. That is, it won't be able to use any class properties unless they have been explicitly set on a partial mock.

If you really wanted to achieve what you're trying to achieve using `passthru` then, you'd need to create a new `setService` method on your action and use it on a partial mock to ensure the class property is initiated.

```php
MyDummyAction::partialMock()
    ->setService($service)
    ->expects('handle')
    ->withArgs([$user])
    ->passthru();
```

Note that this wouldn't work as a non-partial mock since it would complain that there is no expectation defined for the `setService` method.

My advice would be to avoid using `passthru` altogether when possible (with or without Laravel Actions) and instead treat your actions as mockable blocks.

In other words, either test your actions as white box (mocking any underlying layer if necessary like the service for example) or as a black box (mocking the entire action in order to test some higher-level logic).

Even though you've probably sorted out something by now, I hope this helps.
